lib/Target/AVR/AVRISelLowering.cpp

```diff
@@ -1411,14 +1421,16 @@ AVRTargetLowering::LowerReturn(SDValue Chain, CallingConv::ID CallConv,
   // If this is the first return lowered for this function, add the regs to
   // the liveout set for the function.
   MachineFunction &MF = DAG.getMachineFunction();
-  unsigned e = RVLocs.size();
 
-  // Reverse splitted return values to get the "big endian" format required
-  // to agree with the calling convention ABI.
-  ReverseArgumentsToBigEndian(MF, RVLocs);
+  if (RVLocs.size() > 1) {
+    // Reverse splitted return values to get the "big endian" format required
+    // to agree with the calling convention ABI.
+    ReverseReturnValuePartsToEnsureLittleEndian(MF, RVLocs);
+  }
 
   SDValue Flag;
   SmallVector<SDValue, 4> RetOps(1, Chain);
+  unsigned e = RVLocs.size();
```

Comment:
Recommend giving `e` a better name.

Reply:
ok

Reply:
Also, yes, I've had 32 bit values come back from __mulsi3 which were wrong and are now right. That's where my investigation started (avr-rust#130). I also returned these values from my own function and the order is still correct. As you'd expect, the return just does a RET because the values are already in the right registers, r22-r25. So I believe everything is fixed with this patch.